dragonfly/implementations/at_operator.py

```diff
@@ -0,0 +1,472 @@
+except ImportError:
+    pass
+
+import datetime, logging, os, time, json
+# for slack
+import slackclient
+
+from dripline.core import Endpoint, fancy_doc
+from .subprocess_mixin import SlowSubprocessMixin
+
+__all__ = []
+__all__.append('AtOperator')
+
+logger = logging.getLogger(__name__)
+
+@fancy_doc
+class AtOperator(SlowSubprocessMixin, Endpoint):
+    
+    def __init__(self, 
+                 monitor_channel_name = 'slack_test',
+                 update_interval = {"hours":12},
+                 authentication_path = ".project8_authentications.json",
+                 **kwargs):
+        '''
+        monitor_channel_name: the name of Slack monitor channel. 
+        update_interval     : the time interval between regular checks and updates.
+        authentication_path : the absolute path of the file that stores the authentication info.
+        '''
+        imports = ['dateutil', 'funcsigs', 'googleapiclient', 'Credentials']
+        for import_module in imports:
+            if not import_module in globals():
+                raise ImportError(import_module + ' not found, required for AtOperator class.')
+
+        this_home = os.path.expanduser('~')
+        self.authentication_path = os.path.join(this_home, authentication_path)
+        try:
+            json.loads(open(self.authentication_path).read())
+        except IOError, err:
+            logger.critical('The provided authentication file does not exist')
+            os._exit(1)
+        except ValueError, err:
+            logger.critical('The provided authentication file is invalid.')
+            os._exit(1)
+
+        self.slack_client = None
+        self.bot_id = None
+
+        self._creds = None
+
+        self.monitor_channel_name = monitor_channel_name
+        self.monitor_channel_id = ''
+        self.channel_name_to_id_dictionary = {}
+        self.channel_id_to_name_dictionary = {}
+
+        self.current_operator_name = ''
+        self.current_operator_id = ''
+        self.current_shift_end_time = datetime.datetime.now()
+        self.next_shift_start_time = None
+
+        self.temporary_operator_id = []
+
+        self.full_name_to_id_dictionary = {}
+        self.id_to_username_dictionary = {}
+        self.username_to_id_dictionary = {}
+
+        self.command_dictionary = {}
+
+        self.update_interval = datetime.timedelta(**update_interval)
+
+        Endpoint.__init__(self, **kwargs)
+        SlowSubprocessMixin.__init__(self, self.run)
+
+    def get_slack_client(self):
+        '''
+        Set up the slack client.
+        '''
+        slack = ''
+        config_file = json.loads(open(self.authentication_path).read())
+        if 'slack' in config_file and 'operator' in config_file['slack']:
+            slack = config_file['slack']['operator']
+            self.slack_client = slackclient.SlackClient(slack)
+        else:
+            logger.critical('Unable to find slack credentials for operator bot in <~/.project8_authentications.json>')
```

Comment:
since the authentication_path is configurable, this should probably use .format(self.authentication_path)

Reply:
fixed.